presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,308 @@
+
+    public void storeOrMergeSummary(String queryId, String source, int stageId, int taskId, int driverId, DynamicFilterSummary dynamicFilterSummary, int expectedSummariesCount)
+    {
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(sourceDescriptor,
+                ignored -> new DynamicFilterSummaryWithSenders());
+        DynamicFilterSummary mergedSummary = dynamicFilterSummaryWithSenders.addSummary(DynamicFilterSummaryWithSenders.StageTaskKey.of(stageId, taskId), driverId, dynamicFilterSummary, expectedSummariesCount);
+
+        if (mergedSummary != null) {
+            dynamicFilterSummaryWithSenders.getSummaryIfReady()
+                    .ifPresent(summary -> {
+                        SettableFuture<DynamicFilterSummary> future = futures.computeIfAbsent(sourceDescriptor, ignored -> SettableFuture.create());
+                        future.set(summary);
+                    });
+        }
+    }
+
+    /*
+     * registerTasks is required to know whether we have received dynamic filter summaries from all the tasks of query, after which the summary is ready to use
+     * Build side may start reporting summaries through storeOrMergeSummary before registerTasks runs
+     */
+    public void registerTasks(String source, Set<TaskId> taskIds)
+    {
+        if (taskIds.isEmpty()) {
+            return;
+        }
+
+        String queryId = taskIds.iterator().next().getQueryId().getId();
+        log.info("registerTasks source " + source + ", taskIds " + taskIds.stream().map(TaskId::toString).collect(Collectors.joining(",")));
+        checkArgument(taskIds.stream().allMatch(taskId -> taskId.getQueryId().getId().equals(queryId)), "All tasks have to belong to the same query");
+
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, source),
+                ignored -> new DynamicFilterSummaryWithSenders());
+        dynamicFilterSummaryWithSenders.registerTasks(taskIds.stream()
+                .map(DynamicFilterSummaryWithSenders.StageTaskKey::of)
+                .collect(toImmutableSet()));
+    }
+
+    public ListenableFuture<DynamicFilterSummary> getSummary(String queryId, String source)
+    {
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        SettableFuture<DynamicFilterSummary> future = futures.computeIfAbsent(sourceDescriptor, ignored -> SettableFuture.create());
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.get(sourceDescriptor);
+        if (dynamicFilterSummaryWithSenders != null) {
+            dynamicFilterSummaryWithSenders.getSummaryIfReady()
+                    .ifPresent(future::set);
+        }
+        return future;
+    }
+
+    public void removeQuery(String queryId)
```

Comment:
nit: consider using `QueryId` here:
```suggestion
    public void removeQuery(QueryId queryId)
```

Reply:
Done